arangod/Cache/Cache.h

```diff
@@ -274,18 +320,30 @@ class Cache : public std::enable_shared_from_this<Cache> {
   std::size_t const _slotsPerBucket;
 
   // manage eviction rate
-  basics::SharedCounter<64> _insertsTotal;
-  basics::SharedCounter<64> _insertEvictions;
+  struct EvictionStats {
+    basics::SharedCounter<64> insertsTotal;
+    basics::SharedCounter<64> insertEvictions;
+  };
+
+  // this will only flip from false to true, when the eviction stats
+  // are created
+  std::atomic<bool> _haveEvictionStats;
+  // used to create eviction stats under the lock (so that only a single
+  // thread creates them)
+  std::mutex _evictionStatsLock;
+  // this struct is allocated lazily, when the _evictionStats are first written
+  // to
+  std::unique_ptr<EvictionStats> _evictionStats;
```

Comment:
https://github.com/abseil/abseil-cpp/blob/master/absl/base/call_once.h
We could use this to decrease sizeof and speedup it.

Reply:
See above.

Reply:
Well exists easy way to fix it.
Add additional atomic_bool  flag which is set in initial callback.
And make atomic pointer instead of unique_ptr

Second is less convenient. Anyway, both provide better sizeof and better performance than current version

Reply:
I doubt that there is a major performance difference when moving to any of the versions you suggest. Simply because in the general case we only check the value of an atomic variable, which is only once flipped from `false` to `true`. The cost of lazily creating the stats object is there anyway, and it should not matter if we have an extra mutex here or an extra check inside the mutex.
The sizeof values will be better though, but we are talking about a few bytes of out several hundred.

The atomic pointer solution would work, but it has the disadvantage that when multiple threads discover that the pointer is not set, only one of them should create the object. So it needs extra synchronization there.

Reply:
Well suggestion with bool simpler, a little but faster, and have smaller sizeof.

So the question why do we need solution with mutexes?

About atomic pointer I meant different.
Pointer could be used instead of separate bool flag in any solution, so it's just about smaller sizeof and a little more complicated code.


About sizeof, you first who start to optimize it in this PR, also will be good to reorder fields in this class, to make less padding bytes

Reply:
I was trying to optimize things that matter most, which is not the sizeof the Cache object, but the size of the dynamic memory it allocates.

Reply:
> Reduce memory usage for empty in-memory edge caches by ~40%. This is achieved
by allocating each cache's statistics objects only lazily, when actually
needed.

Reply:
Well, in such case I couldn't understand this :(